src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlArrayMethodCallTranslator.cs

```diff
@@ -0,0 +1,408 @@
+
+            case nameof(NpgsqlArrayExtensions.ListFill):
+                return VersionAtLeast(8, 4)
+                    ? new SqlFunctionExpression("array_fill", expression.Method.ReturnType, expression.Arguments.Skip(1))
+                    : throw VersionNotSupportedException(nameof(NpgsqlArrayExtensions.ListFill), 8, 4);
+
+            case nameof(NpgsqlArrayExtensions.MatrixFill):
+                return VersionAtLeast(8, 4)
+                    ? new SqlFunctionExpression("array_fill", expression.Method.ReturnType, expression.Arguments.Skip(1))
+                    : throw VersionNotSupportedException(nameof(NpgsqlArrayExtensions.MatrixFill), 8, 4);
+
+            case nameof(NpgsqlArrayExtensions.ArrayDimensions):
+                return VersionAtLeast(7, 4)
+                    ? new SqlFunctionExpression("array_dims", expression.Method.ReturnType, expression.Arguments.Skip(1).Take(1))
+                    : throw VersionNotSupportedException(nameof(NpgsqlArrayExtensions.ArrayDimensions), 7, 4);
+
+            case nameof(NpgsqlArrayExtensions.ArrayPositions):
+                return VersionAtLeast(9, 5)
+                    ? new SqlFunctionExpression("array_positions", expression.Method.ReturnType, expression.Arguments.Skip(1))
+                    : throw VersionNotSupportedException(nameof(NpgsqlArrayExtensions.ArrayPositions), 9, 5);
+
+            case nameof(NpgsqlArrayExtensions.ArrayRemove):
+                return VersionAtLeast(9, 3)
+                    ? new SqlFunctionExpression("array_remove", expression.Method.ReturnType, expression.Arguments.Skip(1).Take(2))
+                    : throw VersionNotSupportedException(nameof(NpgsqlArrayExtensions.ArrayRemove), 9, 3);
+
+            case nameof(NpgsqlArrayExtensions.ArrayReplace):
+                return VersionAtLeast(9, 3)
+                    ? new SqlFunctionExpression("array_replace", expression.Method.ReturnType, expression.Arguments.Skip(1).Take(3))
+                    : throw VersionNotSupportedException(nameof(NpgsqlArrayExtensions.ArrayReplace), 9, 3);
+
+            case nameof(NpgsqlArrayExtensions.ArrayToString):
+                return VersionAtLeast(9, 1)
+                    ? new SqlFunctionExpression("array_to_string", expression.Method.ReturnType, expression.Arguments.Skip(1).Take(3))
+                    : VersionAtLeast(7, 4)
+                        ? new SqlFunctionExpression("array_to_string", expression.Method.ReturnType, expression.Arguments.Skip(1).Take(2))
+                        : throw VersionNotSupportedException(nameof(NpgsqlArrayExtensions.ArrayToString), 7, 4);
+
+            case nameof(NpgsqlArrayExtensions.StringToArray):
+                return VersionAtLeast(9, 1)
+                    ? new SqlFunctionExpression("string_to_array", expression.Method.ReturnType, expression.Arguments.Skip(1).Take(3))
+                    : VersionAtLeast(7, 4)
```

Comment:
Uhh, as above, supporting pre-9.1 PostgreSQL seems a bit too much... Every bit of code has its price in term of maintenance/complexity etc., at the very least we should wait for users to request support for such old versions rather than eagerly implementing it...

Reply:
Here I was thinking about PostgreSQL-like systems (e.g. Redshift). That said, this is an Npgsql-specific extension method, so anyone seeking it out should know better... I'll rework this to omit 7.4 checks.